liquid-compiler/src/parser.rs

```diff
@@ -3,622 +3,411 @@
 //! This module contains functions than can be used for writing plugins
 //! but should be ignored for simple usage.
 
-use std::collections::HashSet;
-use std::iter::FromIterator;
-use std::slice::Iter;
-
 use liquid_interpreter::Expression;
 use liquid_interpreter::Renderable;
 use liquid_interpreter::Text;
 use liquid_interpreter::Variable;
 use liquid_interpreter::{FilterCall, FilterChain};
+use liquid_value::Scalar;
+use liquid_value::Value;
 
-use super::error::{Error, Result};
-use super::Element;
+use super::error::{/*Error,*/ Result};
 use super::LiquidOptions;
 use super::ParseBlock;
 use super::ParseTag;
-use super::Token;
-
-/// Parses the provided elements into a number of Renderable items
-/// This is the internal version of parse that accepts Elements tokenized
-/// by `lexer::tokenize` and does not register built-in blocks. The main use
-/// for this function is for writing custom blocks.
-///
-/// For parsing from a String you should refer to `liquid::parse`.
-pub fn parse(elements: &[Element], options: &LiquidOptions) -> Result<Vec<Box<Renderable>>> {
-    let mut ret = vec![];
-    let mut iter = elements.iter();
-    let mut token = iter.next();
-    while token.is_some() {
-        let render = match *token.unwrap() {
-            Element::Expression(ref tokens, _) => parse_expression(tokens, options)?,
-            Element::Tag(ref tokens, _) => parse_tag(&mut iter, tokens, options)?,
-            Element::Raw(ref x) => Box::new(Text::new(x.as_str())),
-        };
-        ret.push(render);
-        token = iter.next();
-    }
-    Ok(ret)
-}
 
-const NOTHING: Option<&str> = None;
+use pest::Parser;
 
-pub fn unexpected_token_error<S: ToString>(expected: &str, actual: Option<S>) -> Error {
-    let actual = actual.map(|x| x.to_string());
-    unexpected_token_error_string(expected, actual)
-}
+#[derive(Parser)]
+#[grammar = "grammar.pest"]
+pub struct LiquidParser;
```

Comment:
This is an area that hurts maintainability.  Macros like this make it hard to tell what all is available in the API.

Reply:
At my quick glance at the Pest docs, is Pest all within a trait and this grammar is used for implementing the trait?  So its not like we actually have a bunch of hidden API calls that are created.  Well, I guess the Rules would be.

Reply:
Hmm, I've used `pub struct` but now that I look at it, it seems its not necessary to be `pub`.